core/chains/evm/logpoller/pg_parser.go

```diff
@@ -0,0 +1,226 @@
+	default:
+		panic("unknown sort direction")
+	}
+}
+
+func (v *PgParserVisitor) visitAndFilter(node AndFilter) {
+	for i, filter := range node.filters {
+		if i > 0 {
+			v.query.WriteString(" AND ")
+		}
+		filter.accept(v)
+	}
+}
+
+func (v *PgParserVisitor) visitAddressFilter(node AddressFilter) {
+	switch len(node.address) {
+	case 0:
+		v.errors = append(v.errors, errors.New("address filter must have at least one address signature"))
+	case 1:
+		v.query.WriteString("address = :address")
+		v.args["address"] = node.address[0]
+	default:
+		v.query.WriteString("address = ANY(:address)")
+		v.args["address"] = concatBytes(node.address)
+	}
+}
+
+func (v *PgParserVisitor) visitEventSigFilter(node EventSigFilter) {
+	switch len(node.eventSig) {
+	case 0:
+		v.errors = append(v.errors, errors.New("event sig filter must have at least one event signature"))
+	case 1:
+		v.query.WriteString("event_sig = :event_sig")
+		v.args["event_sig"] = node.eventSig[0].Bytes()
+	default:
+		v.query.WriteString("event_sig = ANY(:event_sig)")
+		v.args["event_sig"] = concatBytes(node.eventSig)
+	}
+}
+
+func (v *PgParserVisitor) visitDataWordFilter(node DataWordFilter) {
+	argumentName, valueName := v.wordValueVariables()
+
+	v.query.WriteString(fmt.Sprintf(
+		"substring(data from 32*:%s+1 for 32) %s :%s",
+		argumentName,
+		node.operator.pgString(),
+		valueName,
+	))
+	v.args[argumentName] = node.index
+	v.args[valueName] = node.value.Bytes()
+}
+
+func (v *PgParserVisitor) visitConfirmationFilter(node ConfirmationFilter) {
+	v.query.WriteString("block_number <= ")
+	v.query.WriteString(nestedBlockNumberQuery(node.confs))
+	v.args["confs"] = node.confs
+}
+
+func (v *PgParserVisitor) visitEvmChainIdFilter(node EvmChainIdFilter) {
+	v.query.WriteString("evm_chain_id = :evm_chain_id")
+	v.args["evm_chain_id"] = node.chainId
+}
+
+func (v *PgParserVisitor) visitTopicFilter(node TopicFilter) {
+	topicIndex, err := sanitizeTopicIndex(node.index)
+	if err != nil {
+		v.errors = append(v.errors, err)
+		return
+	}
+
+	topicIndexName, topicValueName := v.topicVariables()
+	v.query.WriteString(fmt.Sprintf("topics[:%s] %s :%s", topicIndexName, node.operator.pgString(), topicValueName))
+	v.args[topicIndexName] = topicIndex
+	v.args[topicValueName] = node.value.Bytes()
+}
+
+func (v *PgParserVisitor) visitTopicsFilter(node TopicsFilter) {
+	topicIndex, err := sanitizeTopicIndex(node.index)
+	if err != nil {
+		v.errors = append(v.errors, err)
+		return
+	}
+
+	topicIndexName, topicValueName := v.topicVariables()
+	v.query.WriteString(fmt.Sprintf("topics[:%s] = ANY(:%s)", topicIndexName, topicValueName))
+	v.args[topicIndexName] = topicIndex
+	v.args[topicValueName] = concatBytes(node.values)
+}
+
+func (v *PgParserVisitor) visitNamedFilter(node NamedFilter) {
+	v.filterNameIndex++
+	argumentValue := fmt.Sprintf("%s_%d", node.fieldName, v.filterNameIndex)
+
+	v.query.WriteString(fmt.Sprintf("%s %s :%s", node.fieldName, node.operator.pgString(), argumentValue))
```

Comment:
> to make sure it doesn't contain any tricky characters like ', ", whitespace, etc. 

Actually, I guess we don't need to exclude those we just need to call something that escapes them properly to ensure it gets interpreted as a single column identifier.

Alternatively, do we really need to expose `NamedFilter` to the outside, or can we have that be something that's hidden away so it can only be used internally to build other filters not as a filter in itself?

Reply:
Agree, exposing NamedFilter might be too risky, I've added it to reduce code duplication, but I haven't thought about the security concerns - I will verify solution without `NamedFilter`. Regarding sql injection, all these queries are passed to `SelectNamed`\`BindNamed` which should work as sql injection protection layer.

Reply:
@reductionista I've removed NamedFilter. API should be more defensive now and allow passing only known/predefined filters.